rfcs/0144-versioned-flake-references.md

```diff
@@ -0,0 +1,118 @@
+have to be a valid version defiined by the versioning scheme, `1` and `1.0` are
+also valid versions for version requirements. The comparator has to be one of
+the following options:
+- `^` compatible (as defined by the versioning scheme)
+- `<` less than
+- `<=` less than or equal to
+- `>` greater than
+- `>=` greater than or equal to
+- `=` equal to
+
+A version placeholder with no version requirements will match all valid
+versions defined by the versioning scheme.
+
+## Versioning scheme
+The versions will follow [semantic versioning] (semver). The tags have to
+follow semver, `1.0` and `1.0.0.0` will not match the placeholder `{=1.0}` as
+it is not a valid version defined by semver. Build identifiers will not be
+taken into consideration when calculating version requirements, and the latest
+tag will be selected if multiple tags have the same precedence defined by
+semver.
+
+## Upgrading
+A new Nix command, `nix flake upgrade`, will edit `flake.nix` and upgrade all
+the version placeholders in the inputs. A new flag, `--upgrade-input` will
+upgrade only the specified flake input. Only version placeholders that contain
+exactly one version requirement which the comparator is either `^` or `=` will
+be upgraded. All other version placeholders will be left unchanged.
+
+# Examples and Interactions
+[examples-and-interactions]: #examples-and-interactions
+
+Here are the git tags of a hypothetical Nix flake "github:foo/bar":
+- v0.1.0
+- v0.1.1
+- v0.2.0
+- v1.0.0
+- v1.0.1
+- v1.1.0
+- v2.0.0
+- unrelated
+
+Flake reference | The tag it points to | The `upgrade`ed flake reference
+-|-|-
+`github:foo/bar/v{^1}` | `v1.1.0` | `github:foo/bar/v{^2}`
+`github:foo/bar/v{=1.0}` | `v1.0.1` | `github:foo/bar/v{=2.0}`
+`github:foo/bar/v{^1.0}` | `v1.1.0` | `github:foo/bar/v{^2.0}`
+`github:foo/bar/v{=1.0.0}` | `v1.0.0` | `github:foo/bar/v{=2.0.0}`
+`github:foo/bar/v{^1.0.0}` | `v1.1.0` | `github:foo/bar/v{^2.0.0}`
+`github:foo/bar/v{^0.1.0}` | `v0.1.1` | `github:foo/bar/v{^2.0.0}`
+`github:foo/bar/v{<2}` | `v1.1.0` | `github:foo/bar/v{<2}` (no change)
+`github:foo/bar/v{>1.0}` | `v2.0.0` | `github:foo/bar/v{>1.0}` (no change)
+`github:foo/bar/v{^1,<1.1}` | `v1.0.1` | `github:foo/bar/v{^1,<1.1}` (no change)
+`github:foo/bar/v{}` | `v2.0.0` | `github:foo/bar/v{}` (no change)
+
+# Drawbacks
+[drawbacks]: #drawbacks
+
+- This is impossible with some types of flake references, such as `path`s and
+  `git+` URLs, which will make the flake interface less consistent.
+- This adds extra complexity to `nix flake update`.
+- The difference between `update` and `upgrade` might cause confusion.
+
+# Alternatives
+[alternatives]: #alternatives
+
+- Using a different syntax for the placeholders
+- Using a version scheme other than semantic versioning
+- Adding flags to `nix flake update` instead of creating a new `upgrade` command
+- Not adding the feature, manually update the flake references
+
+# Unresolved questions
+[unresolved]: #unresolved-questions
```

Comment:
Not discussed here is the algorithm for selecting versions. In conjunction with follows/overrides, finding a feasible solution could become expensive, e.g. in
```
inputs.foo.url = "github:bla/foo/{>1.2,<1.5}";
inputs.bar.url = "github:bla/bar/{^1}";
inputs.bar.inputs.foo.follows = "foo";
```
Nix would need to find a version of `bar` whose `foo` input has constraints compatible with the ones given here.

I wouldn't want to have a SAT solver in Nix...

Reply:
> I wouldn't want to have a SAT solver in Nix...

We should consider having one outside of Nix, making lockfile generation pluggable.
Perhaps the version constraints should be in a separate attribute that's open for extension?

In fact, we could consider the lock reading and fetching to be part of Nix, and the Flake CLI to be one implementation of the broader Nix package management idea. Dream2nix is also in the business of lock file generation. Perhaps the role of Nix should be limited to the *integration* of lock file generators, and not actually implementing them. We don't even need a dumb lock file generator like we currently have because that one could just as well live in a flake that's already been locked.

Nix should focus on its core competencies that it does very well.
 - building immutable store paths
 - bootstrapping other software

*Anything* that's built on top of that functionality is going to feel as good as native, plus have the benefit of pinning, patching, and decentralized development.

Flakes is oddly centralized for a package manager built on Nix...

Reply:
Late to the party, but let me just write this down just in case. One way to avoid the need for SAT-solving is to allow only semver open (`^1.2.3`) version requirements. That way, a greedy algorithm which just always picks the latest semver-compatible version works. 

Moreover, in practice, in Cargo:

* 99% of constraints are in fact just `^1.2.3` (it was _genius_ design decision to make `version="1.2.3"` be a syntactic sugar for `version="^1.2.3"`)
* people often try to use non-semver open constraints, most of the time that's a wrong thing to do, and leads to problem down the line (eg, their build works, but builds of folks depending on them gridlocks due to unsatisfiable constraints)
* at some point we even considered deprecating non-`^` constraints, but didn't do that, because there are occasionally good reasons for them

So, perhaps a safe move wold be to allow only `^1.2.3` constraints, (and using only implicit `1.2.3` syntax). If that turns out to be expressive enough, that would be a massive simplification. If not, fancier constrains can be added later.